pkg/attribute/mutableBag.go

```diff
@@ -96,10 +111,58 @@ func copyValue(v interface{}) interface{} {
 	return v
 }
 
-// Done indicates the bag can be reclaimed.
-func (mb *MutableBag) Done() {
+// Child allocates a derived mutable bag.
+func (mb *MutableBag) Child() *MutableBag {
+	return GetMutableBag(mb)
+}
+
+// childDone indicates that a child bag has been reclaimed.
+func (mb *MutableBag) childDone() {
+	mb.Lock()
+	mb.children--
+	// if this bag was already scheduled for done()
+	// then perform the operation.
+	if mb.doneScheduled && mb.children == 0 {
```

Comment:
I don't like the complexity introduced by this 'doneScheduled' feature. 

All we're trying to do here is prevent misuse of the API and not to make extra affordances for these bad uses. Instead of ever setting doneScheduled to true, let's just panic at that point: "hey dummy, you're marking a parent bag done while there are still outstanding children". This shouldn't happen in a running system.